File: flutter/circlebox_flutter/android/src/main/kotlin/com/circlebox/flutter_circlebox/CircleboxFlutterPlugin.kt
package com.circlebox.flutter_circlebox

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File

/**
 * Flutter bridge for the Android CircleBox SDK.
 *
 * Reflection is used so the plugin can compile even when the host app has not
 * wired the native CircleBox dependency yet.
 */
class CircleboxFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "circlebox_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                "start" -> {
                    @Suppress("UNCHECKED_CAST")
                    val config = call.arguments as? Map<String, Any?>
                    invokeStart(config)
                    result.success(null)
                }

                "breadcrumb" -> {
                    val message = call.argument<String>("message") ?: ""
                    @Suppress("UNCHECKED_CAST")
                    val attrs = call.argument<Map<String, String>>("attrs") ?: emptyMap()
                    invokeBreadcrumb(message, attrs)
                    result.success(null)
                }

                "exportLogs" -> {
                    @Suppress("UNCHECKED_CAST")
                    val args = call.arguments as? Map<String, Any?>
                    @Suppress("UNCHECKED_CAST")
                    val formats = args?.get("formats") as? List<String> ?: emptyList()
                    val files = if (formats.isEmpty()) invokeExportLogs() else invokeExportLogs(formats)
                    result.success(files)
                }

                "hasPendingCrashReport" -> {
                    val pending = invokeNoArg("hasPendingCrashReport") as? Boolean ?: false
                    result.success(pending)
                }

                "clearPendingCrashReport" -> {
                    invokeNoArg("clearPendingCrashReport")
                    result.success(null)
                }

                "debugSnapshot" -> {
                    @Suppress("UNCHECKED_CAST")
                    val args = call.arguments as? Map<String, Any?>
                    val maxEvents = (args?.get("maxEvents") as? Number)?.toInt() ?: 200
                    val events = invokeDebugSnapshot(maxEvents)
                    result.success(events)
                }

                else -> result.notImplemented()
            }
        } catch (error: ClassNotFoundException) {
            result.error("missing_native_sdk", "Native CircleBox Android SDK not found in host app", error.message)
        } catch (error: Throwable) {
            result.error("circlebox_error", error.message, null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun invokeNoArg(methodName: String): Any? {
        val clazz = Class.forName("com.circlebox.sdk.CircleBox")
        val method = clazz.getMethod(methodName)
        return method.invoke(null)
    }

    private fun invokeStart(config: Map<String, Any?>?) {
        val clazz = Class.forName("com.circlebox.sdk.CircleBox")
        if (config == null) {
            invokeNoArg("start")
            return
        }

        val configClass = Class.forName("com.circlebox.sdk.CircleBoxConfig")
        val bufferCapacity = (config["bufferCapacity"] as? Number)?.toInt() ?: 50
        val jankThresholdMs = (config["jankThresholdMs"] as? Number)?.toLong() ?: 200L
        val sanitizeAttributes = (config["sanitizeAttributes"] as? Boolean) ?: true
        val maxAttributeLength = (config["maxAttributeLength"] as? Number)?.toInt() ?: 256
        val diskCheckIntervalSec = (config["diskCheckIntervalSec"] as? Number)?.toLong() ?: 60L
        val enableDebugViewer = (config["enableDebugViewer"] as? Boolean) ?: false

        val instance = runCatching {
            val ctor = configClass.getConstructor(
                Int::class.javaPrimitiveType,
                Long::class.javaPrimitiveType,
                Boolean::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                Long::class.javaPrimitiveType,
                Boolean::class.javaPrimitiveType
            )
            ctor.newInstance(
                bufferCapacity,
                jankThresholdMs,
                sanitizeAttributes,
                maxAttributeLength,
                diskCheckIntervalSec,
                enableDebugViewer
            )
        }.getOrElse {
            val ctor = configClass.getConstructor(
                Int::class.javaPrimitiveType,
                Long::class.javaPrimitiveType,
                Boolean::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                Long::class.javaPrimitiveType
            )
            ctor.newInstance(
                bufferCapacity,
                jankThresholdMs,
                sanitizeAttributes,
                maxAttributeLength,
                diskCheckIntervalSec
            )
        }

        val method = clazz.getMethod("start", configClass)
        method.invoke(null, instance)
    }

    @Suppress("UNCHECKED_CAST")
    private fun invokeBreadcrumb(message: String, attrs: Map<String, String>) {
        val clazz = Class.forName("com.circlebox.sdk.CircleBox")
        val withAttrs = runCatching { clazz.getMethod("breadcrumb", String::class.java, Map::class.java) }.getOrNull()
        if (withAttrs != null) {
            withAttrs.invoke(null, message, attrs)
            return
        }

        val withoutAttrs = clazz.getMethod("breadcrumb", String::class.java)
        withoutAttrs.invoke(null, message)
    }

    private fun invokeExportLogs(): List<String> {
        val raw = invokeNoArg("exportLogs")
        return mapFiles(raw)
    }

    @Suppress("UNCHECKED_CAST")
    private fun invokeExportLogs(formats: List<String>): List<String> {
        val clazz = Class.forName("com.circlebox.sdk.CircleBox")
        val exportEnumClass = Class.forName("com.circlebox.sdk.CircleBoxExportFormat")
        val set = LinkedHashSet<Any>()
        formats.forEach { item ->
            val enumName = item.uppercase()
            runCatching {
                val value = exportEnumClass.getMethod("valueOf", String::class.java).invoke(null, enumName)
                set.add(value)
            }
        }

        if (set.isEmpty()) {
            return invokeExportLogs()
        }

        val method = clazz.getMethod("exportLogs", Set::class.java)
        val raw = method.invoke(null, set)
        return mapFiles(raw)
    }

    private fun invokeDebugSnapshot(maxEvents: Int): List<Map<String, Any?>> {
        val clazz = Class.forName("com.circlebox.sdk.CircleBox")
        val method = clazz.getMethod("debugSnapshot", Int::class.javaPrimitiveType)
        val raw = method.invoke(null, maxEvents) as? List<*> ?: return emptyList()
        return raw.mapNotNull { event ->
            val instance = event ?: return@mapNotNull null
            val eventClass = instance.javaClass

            val attrsRaw = runCatching {
                @Suppress("UNCHECKED_CAST")
                eventClass.getMethod("getAttrs").invoke(instance) as? Map<Any?, Any?>
            }.getOrNull() ?: emptyMap()
            val attrs = LinkedHashMap<String, String>(attrsRaw.size)
            attrsRaw.forEach { (key, value) ->
                attrs[key.toString()] = value?.toString() ?: ""
            }

            linkedMapOf(
                "seq" to runCatching { eventClass.getMethod("getSeq").invoke(instance) as? Number }.getOrNull()?.toLong(),
                "timestamp_unix_ms" to runCatching { eventClass.getMethod("getTimestampUnixMs").invoke(instance) as? Number }.getOrNull()?.toLong(),
                "uptime_ms" to runCatching { eventClass.getMethod("getUptimeMs").invoke(instance) as? Number }.getOrNull()?.toLong(),
                "type" to runCatching { eventClass.getMethod("getType").invoke(instance)?.toString() }.getOrNull(),
                "thread" to runCatching { eventClass.getMethod("getThread").invoke(instance)?.toString()?.lowercase() }.getOrNull(),
                "severity" to runCatching { eventClass.getMethod("getSeverity").invoke(instance)?.toString()?.lowercase() }.getOrNull(),
                "attrs" to attrs
            )
        }
    }

    private fun mapFiles(raw: Any?): List<String> {
        val files = raw as? List<*> ?: return emptyList()
        return files.mapNotNull { item ->
            when (item) {
                is File -> item.absolutePath
                is String -> item
                else -> null
            }
        }
    }
}

File: flutter/circlebox_flutter/ios/Classes/CircleboxFlutterPlugin.swift
import Flutter
import UIKit

#if canImport(CircleBoxSDK)
import CircleBoxSDK
#endif

/// Flutter bridge for the iOS CircleBox SDK.
public class CircleboxFlutterPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "circlebox_flutter", binaryMessenger: registrar.messenger())
    let instance = CircleboxFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "start":
      #if canImport(CircleBoxSDK)
      let args = call.arguments as? [String: Any] ?? [:]
      let config = CircleBoxConfig(
        bufferCapacity: args["bufferCapacity"] as? Int ?? 50,
        jankThresholdMs: UInt64(args["jankThresholdMs"] as? Int ?? 200),
        sanitizeAttributes: args["sanitizeAttributes"] as? Bool ?? true,
        maxAttributeLength: args["maxAttributeLength"] as? Int ?? 256,
        diskCheckIntervalSec: args["diskCheckIntervalSec"] as? TimeInterval ?? 60,
        enableDebugViewer: args["enableDebugViewer"] as? Bool ?? false
      )
      CircleBox.start(config: config)
      result(nil)
      #else
      // Keep plugin buildable without forcing every host app to link CircleBoxSDK immediately.
      result(FlutterError(code: "missing_native_sdk", message: "CircleBoxSDK is not linked in the iOS host app", details: nil))
      #endif

    case "breadcrumb":
      guard
        let args = call.arguments as? [String: Any],
        let message = args["message"] as? String
      else {
        result(FlutterError(code: "bad_args", message: "Missing message", details: nil))
        return
      }

      let attrs = args["attrs"] as? [String: String] ?? [:]

      #if canImport(CircleBoxSDK)
      CircleBox.breadcrumb(message, attrs: attrs)
      result(nil)
      #else
      result(FlutterError(code: "missing_native_sdk", message: "CircleBoxSDK is not linked in the iOS host app", details: nil))
      #endif

    case "exportLogs":
      #if canImport(CircleBoxSDK)
      do {
        let args = call.arguments as? [String: Any] ?? [:]
        let rawFormats = args["formats"] as? [String] ?? ["json", "csv"]
        var formats = Set<CircleBoxExportFormat>()
        for item in rawFormats {
          switch item.lowercased() {
          case "json":
            formats.insert(.json)
          case "csv":
            formats.insert(.csv)
          case "json_gzip":
            formats.insert(.jsonGzip)
          case "csv_gzip":
            formats.insert(.csvGzip)
          case "summary":
            formats.insert(.summary)
          default:
            continue
          }
        }
        if formats.isEmpty {
          formats = [.json, .csv]
        }

        let urls = try CircleBox.exportLogs(formats: formats)
        result(urls.map { $0.path })
      } catch {
        result(FlutterError(code: "export_failed", message: error.localizedDescription, details: nil))
      }
      #else
      result(FlutterError(code: "missing_native_sdk", message: "CircleBoxSDK is not linked in the iOS host app", details: nil))
      #endif

    case "hasPendingCrashReport":
      #if canImport(CircleBoxSDK)
      result(CircleBox.hasPendingCrashReport())
      #else
      result(false)
      #endif

    case "clearPendingCrashReport":
      #if canImport(CircleBoxSDK)
      do {
        try CircleBox.clearPendingCrashReport()
        result(nil)
      } catch {
        result(FlutterError(code: "clear_failed", message: error.localizedDescription, details: nil))
      }
      #else
      result(FlutterError(code: "missing_native_sdk", message: "CircleBoxSDK is not linked in the iOS host app", details: nil))
      #endif

    case "debugSnapshot":
      #if canImport(CircleBoxSDK)
      let args = call.arguments as? [String: Any] ?? [:]
      let maxEvents = args["maxEvents"] as? Int ?? 200
      let events = CircleBox.debugSnapshot(maxEvents: maxEvents).map { event in
        [
          "seq": event.seq,
          "timestamp_unix_ms": event.timestampUnixMs,
          "uptime_ms": event.uptimeMs,
          "type": event.type,
          "thread": event.thread.rawValue,
          "severity": event.severity.rawValue,
          "attrs": event.attrs
        ] as [String: Any]
      }
      result(events)
      #else
      result([])
      #endif

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
